extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/dart/controller/DartControllerContext.java

```diff
@@ -0,0 +1,245 @@
+  /**
+   * Default for {@link ControllerQueryKernelConfig#getMaxConcurrentStages()}.
+   */
+  public static final int DEFAULT_MAX_CONCURRENT_STAGES = 2;
+
+  /**
+   * Default for {@link MultiStageQueryContext#getTargetPartitionsPerWorkerWithDefault(QueryContext, int)}.
+   */
+  public static final int DEFAULT_TARGET_PARTITIONS_PER_WORKER = 1;
+
+  /**
+   * Context parameter for maximum number of nonleaf workers.
+   */
+  public static final String CTX_MAX_NON_LEAF_WORKER_COUNT = "maxNonLeafWorkers";
+
+  /**
+   * Default to scatter/gather style: fan in to a single worker after the leaf stage(s).
+   */
+  public static final int DEFAULT_MAX_NON_LEAF_WORKER_COUNT = 1;
+
+  private final Injector injector;
+  private final ObjectMapper jsonMapper;
+  private final ObjectMapper smileMapper;
+  private final DruidNode selfNode;
+  private final ControllerServerId selfId;
+  private final ServiceClientFactory serviceClientFactory;
+  private final BrokerServerView serverView;
+  private final MemoryIntrospector memoryIntrospector;
+  private final ServiceMetricEvent.Builder metricBuilder;
+  private final ServiceEmitter emitter;
+
+  @Inject
+  public DartControllerContext(
+      final Injector injector,
+      @Json final ObjectMapper jsonMapper,
+      @Smile final ObjectMapper smileMapper,
+      @Self final DruidNode selfNode,
+      final ControllerServerId selfId,
+      @EscalatedGlobal final ServiceClientFactory serviceClientFactory,
+      final MemoryIntrospector memoryIntrospector,
+      final BrokerServerView serverView,
+      final ServiceEmitter emitter
+  )
+  {
+    this.injector = injector;
+    this.jsonMapper = jsonMapper;
+    this.smileMapper = smileMapper;
+    this.selfNode = selfNode;
+    this.selfId = selfId;
+    this.serviceClientFactory = serviceClientFactory;
+    this.serverView = serverView;
+    this.memoryIntrospector = memoryIntrospector;
+    this.metricBuilder = new ServiceMetricEvent.Builder();
+    this.emitter = emitter;
+  }
+
+  @Override
+  public ControllerQueryKernelConfig queryKernelConfig(
+      final String queryId,
+      final MSQSpec querySpec
+  )
+  {
+    final List<DruidServerMetadata> servers = serverView.getDruidServerMetadatas();
+
+    // Lock in the list of workers when creating the kernel config. There is a race here: the serverView itself is
+    // allowed to float. If a segment moves to a new server that isn't part of our list after the WorkerManager is
+    // created, we won't be able to find a valid server for certain segments. This isn't expected to be a problem,
+    // since the serverView is referenced shortly after the worker list is created.
+    final List<String> workerIds = new ArrayList<>(servers.size());
+    for (final DruidServerMetadata server : servers) {
+      workerIds.add(WorkerId.fromDruidServerMetadata(server, queryId).toString());
+    }
+
+    // Shuffle workerIds, so we don't bias a single server to always be worker 0 (which tends to do more work).
```

Comment:
Expanded the comment to say why:

> Shuffle workerIds, so we don't bias towards specific servers when running multiple queries concurrently. For any
given query, lower-numbered workers tend to do more work, because the controller prefers using lower-numbered
workers when maxWorkerCount for a stage is less than the total number of workers.